### PR TITLE
[Storage] [Typing] [Blobs] Final Typing Sweep

### PR DESCRIPTION
**Pull request by vincenttran-msft**
GREEN MYPY 🕺 
image

### CODE REVIEW

File: sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py [hunk 1]
```diff
@@ -614,7 +614,7 @@ def download_blob(
         self, offset: Optional[int] = None,
         length: Optional[int] = None,
         *,
-        encoding: Optional[str] = None,
+        encoding: None = None,
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py:617]: I think this was the issue with our overload. If my understanding is correct, if encoding is str --> will return a `StorageStreamDownloader[str]`. If encoding is unset, will be `StorageStreamDownloader[bytes]`.

So our issue was that previously we had the bytes case have an `encoding: Optional[str]` type-hint, which covers the domain of regular `str`. Thus the overload wasn't able to "gate" properly and know which one should result to which return type. 

BUT I saw we didn't have this issue in one of the container client's, and that cleared everything up 😄

File: sdk/storage/azure-storage-blob/azure/storage/blob/_deserialize.py [hunk 2]
```diff
@@ -160,7 +160,7 @@ def get_blob_properties_from_generated_code(generated: "BlobItemInternal") -> Bl
     blob.deleted = generated.deleted
     blob.snapshot = generated.snapshot
     blob.is_append_blob_sealed = generated.properties.is_sealed
-    blob.metadata = generated.metadata.additional_properties if generated.metadata else {}
+    blob.metadata = generated.metadata.additional_properties if generated.metadata else {}  # type: ignore [assignment]
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_deserialize.py:163]: Generated's return type here is too permissive, so just throwing an ignore.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_lease.py [hunk 3]
```diff
@@ -42,7 +41,6 @@ class BlobLeaseClient(): # pylint: disable=client-accepts-api-version-keyword
     last_modified: Optional["datetime"]
     """The last modified timestamp of the lease currently being maintained.
     This will be `None` if no lease has yet been acquired or modified."""
-    _client: Union["BlobOperations", "ContainerOperations"]
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_lease.py:45]: I figured it out. This is the only client where we annotated the internal `_client`, and thus MyPy was too smart for it's own good and hurt itself in confusion. By omitting this, this makes the generated code call undiscoverable, thus not enforcing type checking on the return 😄

File: sdk/storage/azure-storage-blob/azure/storage/blob/aio/_upload_helpers.py [hunk 4]
```diff
@@ -67,18 +70,20 @@ async def upload_block_blob(  # pylint: disable=too-many-locals, too-many-statem
...
             if encryption_options.get('key'):
+                if not isinstance(data, bytes):
+                    raise TypeError('Blob data should be of type bytes.')
                 encryption_data, data = encrypt_blob(data, encryption_options['key'], encryption_options['version'])
                 headers['x-ms-meta-encryptiondata'] = encryption_data
 
-            response = await client.upload(
-                body=data,
+            response = cast(Dict[str, Any], await client.upload(
+                body=data,  # type: ignore [arg-type]
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/aio/_upload_helpers.py:86]: Curious what the error is here? Like what type it wants and what it gets.

**Review by jalauzon-msft — APPROVED** [review]

**Pull request merged**